share/p2p/shrexnd/client.go

```diff
@@ -0,0 +1,173 @@
+package shrexnd
+
+import (
+	"errors"
+	"fmt"
+
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"golang.org/x/net/context"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	share_p2p_v1 "github.com/celestiaorg/celestia-node/share/p2p/shrexnd/v1/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+var (
+	errNotAvailable = errors.New("none of the peers has requested data")
+)
+
+// Client implements Getter interface, requesting data from remote peers
+type Client struct {
+	host           host.Host
+	defaultTimeout time.Duration
+
+	// fake peers for testing purposes
+	testPeers []peer.ID
+}
+
+// NewClient creates a new shrEx client
+func NewClient(host host.Host, defaultTimeout time.Duration) *Client {
+	return &Client{
+		host:           host,
+		defaultTimeout: defaultTimeout,
+	}
+}
+
+// GetSharesByNamespace request shares with option to collect proofs from remote peers using shrex
+// protocol
+func (c *Client) GetSharesByNamespace(
+	ctx context.Context,
+	root *share.Root,
+	nID namespace.ID,
+	peerIDs peer.IDSlice,
+) (share.NamespacedShares, error) {
+	// overwrite peers in test
+	if len(c.testPeers) != 0 {
+		peerIDs = c.testPeers
+	}
+
+	for _, peer := range peerIDs {
```

Comment:
Should it retry until it receives a valid response? It only requests once per peer here. I think this isn't clear in the ADR

Reply:
Disscussed with @distractedm1nd  and @Wondertan  to move retry logic out of client and let the caller decide on retry